src/Analysis/Engine/Test/LineFormatterTests.cs

```diff
@@ -0,0 +1,460 @@
+            AssertSingleLineFormat("{*range(4), 4}");
+            AssertSingleLineFormat("{'x': 1, **{'y': 2}}");
+            AssertSingleLineFormat("{'x': 1, **{'x': 2}}");
+            AssertSingleLineFormat("{**{'x': 2}, 'x': 1}");
+        }
+
+        [TestMethod, Priority(0)]
+        public void MultilineStringAssignment() {
+            AssertSingleLineFormat("x='''abc\ntest'''abc", "x = '''abc");
+        }
+
+        [TestMethod, Priority(0)]
+        public void MultilineDefaultArg() {
+            AssertSingleLineFormat("def foo(x='''abc\ntest''')", "def foo(x='''abc");
+        }
+
+        [TestMethod, Priority(0)]
+        public void LineContinuation() {
+            AssertSingleLineFormat("a+b+ \\\n", "a + b + \\");
+        }
+
+        [TestMethod, Priority(0)]
+        public void MultilineChainedCall() {
+            const string code = "foo.a() \\\n   .b() \\\n   .c()";
+            AssertSingleLineFormat(code, "foo.a() \\");
+            AssertSingleLineFormat(code, ".b() \\", line: 2, editStart: 4);
+            AssertSingleLineFormat(code, ".c()", line: 3, editStart: 4);
+        }
+
+        [TestMethod, Priority(0)]
+        public void BracketCommas() {
+            AssertSingleLineFormat("a[:, :, :, 1]");
+            AssertSingleLineFormat("a[x:y, x + 1 :y, :, 1]");
+            AssertSingleLineFormat("a[:, 1:3]");
+            AssertSingleLineFormat("a[:, :3, :]");
+            AssertSingleLineFormat("a[:, 3:, :]");
+        }
+
+        [TestMethod, Priority(0)]
+        public void MultilineStringTrailingComment() {
+            AssertSingleLineFormat(@"'''
+foo
+''' # comment", "  # comment", line: 3, editStart: 4);
+        }
+
+        [TestMethod, Priority(0)]
+        public void Backtick() {
+            AssertSingleLineFormat("`a`", languageVersion: PythonLanguageVersion.V27);
+            AssertSingleLineFormat("foo(`a`)", languageVersion: PythonLanguageVersion.V27);
+            AssertSingleLineFormat("`a` if a else 'oops'", languageVersion: PythonLanguageVersion.V27);
+        }
+
+        [TestMethod, Priority(0)]
+        public void ExecStatement() {
+            AssertSingleLineFormat("exec code", languageVersion: PythonLanguageVersion.V27);
+            AssertSingleLineFormat("exec (code)", languageVersion: PythonLanguageVersion.V27);
+            AssertSingleLineFormat("exec(code)");
+        }
+
+        [TestMethod, Priority(0)]
+        public void GrammarFile() {
+            var src = TestData.GetPath("TestData", "Formatting", "pythonGrammar.py");
+
+            string fileContents;
+            using (var reader = new StreamReader(src, true)) {
+                fileContents = reader.ReadToEnd();
+            }
+
+            var lines = fileContents.Split(new[] { "\r\n", "\r", "\n" }, StringSplitOptions.None);
+
+            using (var reader = new StringReader(fileContents)) {
+                var lineFormatter = new LineFormatter(reader, PythonLanguageVersion.V37);
```

Comment:
Can we use `PythonVersions.LatestAvailable3X`, or that code works specifically with `V37`?

Reply:
There's nothing special about V37, but I need a `PythonLanguageVersion` (the enum). LatestAvailable3X is an `InterpretorConfiguration` which does not contain an enum value.

Reply:
Ok, do you actually need python to be installed for these tests?

Reply:
AFAIK this setting is to tokenizer to know which language syntax to use

Reply:
Ok, ignore the comment.